chain/store/store.go

```diff
@@ -730,38 +737,82 @@ func (cs *ChainStore) MessagesForTipset(ts *types.TipSet) ([]types.ChainMsg, err
 		return nil
 	}
 
-	var out []types.ChainMsg
+	selectMsg := func(m *types.Message) (bool, error) {
+		if err := preloadAddr(m.From); err != nil {
+			return false, err
+		}
+
+		if applied[m.From] != m.Nonce {
+			return false, nil
+		}
+		applied[m.From]++
+
+		// TODO: We should be able to take this out -- such a message can never enter mempool or a valid block
+		if balances[m.From].LessThan(m.RequiredFunds()) {
+			return false, nil
+		}
+		balances[m.From] = types.BigSub(balances[m.From], m.RequiredFunds())
+
+		return true, nil
+	}
+
+	var out []BlockMessages
 	for _, b := range ts.Blocks() {
+
 		bms, sms, err := cs.MessagesForBlock(b)
 		if err != nil {
 			return nil, xerrors.Errorf("failed to get messages for block: %w", err)
 		}
 
-		cmsgs := make([]types.ChainMsg, 0, len(bms)+len(sms))
-		for _, m := range bms {
-			cmsgs = append(cmsgs, m)
-		}
-		for _, sm := range sms {
-			cmsgs = append(cmsgs, sm)
+		bm := BlockMessages{
+			Miner:         b.Miner,
+			BlsMessages:   make([]types.ChainMsg, 0, len(bms)),
+			SecpkMessages: make([]types.ChainMsg, 0, len(sms)),
+			WinCount:      b.ElectionProof.WinCount,
 		}
 
-		for _, cm := range cmsgs {
-			m := cm.VMMessage()
-			if err := preloadAddr(m.From); err != nil {
-				return nil, err
+		for _, bmsg := range bms {
+			b, err := selectMsg(bmsg.VMMessage())
+			if err != nil {
+				return nil, xerrors.Errorf("failed to decide whether to select message for block: %w", err)
+			}
+
+			if b {
+				bm.BlsMessages = append(bm.BlsMessages, bmsg)
 			}
+		}
 
-			if applied[m.From] != m.Nonce {
-				continue
+		for _, smsg := range sms {
+			b, err := selectMsg(smsg.VMMessage())
+			if err != nil {
+				return nil, xerrors.Errorf("failed to decide whether to select message for block: %w", err)
 			}
-			applied[m.From]++
 
-			if balances[m.From].LessThan(m.RequiredFunds()) {
-				continue
+			if b {
+				bm.SecpkMessages = append(bm.SecpkMessages, smsg)
 			}
-			balances[m.From] = types.BigSub(balances[m.From], m.RequiredFunds())
+		}
+
+		out = append(out, bm)
+	}
+
+	return out, nil
+}
+
+func (cs *ChainStore) MessagesForTipset(ts *types.TipSet) ([]types.ChainMsg, error) {
```

Comment:
This method is now making 2 passes (the one in `BlockMsgsForTipset`, and then the one in here to flatten out the result into []ChainMsg). It's not great, but probably only slows down stuff like searching for messages, estimating gas premium, etc. 

Can be improved.